include/small/obuf_asan.h

```diff
@@ -0,0 +1,200 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <sys/uio.h>
+#include <stdbool.h>
+
+#include "slab_cache.h"
+#include "util.h"
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+enum {
+	/**
+	 * Maximum index in iovec vectors list. Note that the list is
+	 * NULL terminated thus maximum index holding vector is one less.
+	 */
+	SMALL_OBUF_IOV_MAX = IOV_MAX,
+	/**
+	 * Number of exponentially growing vectors at the end of iovec
+	 * vectors list.
+	 */
+	SMALL_OBUF_IOV_GEOMETRIC_SIZE = 32,
+	/**
+	 * Number of vectors at the beginning of iovec vector list
+	 * that are allocated with distinct malloc call. Let's refer
+	 * them as checked vectors.
+	 */
+	SMALL_OBUF_IOV_CHECKED_SIZE = SMALL_OBUF_IOV_MAX + 1 -
+				      SMALL_OBUF_IOV_GEOMETRIC_SIZE,
+	/** Alignment for the checked vectors. */
+	SMALL_OBUF_ALIGNMENT = 1,
+	/** Minimum size when reserving memory. See struct obuf comment. */
+	SMALL_OBUF_MIN_RESERVE = 128,
+};
+
+/**
+ * ASAN friendly implementation for obuf allocator. It has same interface as
+ * regular implementation but at the beginning it uses malloc() for every
+ * allocation. More precisely for the first (in terms of obuf_iovcnt())
+ * SMALL_OBUF_IOV_CHECKED_SIZE allocations. Then the memory is allocated in
+ * exponentially growing blocks which then used to provide requested
+ * allocations as in the regular implementation.
+ *
+ * This combined strategy allows to do out-of-bound access checks for first
+ * allocations on the one hand and on the other hand limit the size of iov
+ * vector list. The latter is required to avoid iov vector list reallocations
+ * which is expected by some client code.
+ *
+ * See however a bit of limitation for out-of-bound access check in description
+ * of small_asan_alloc.
+ *
+ * Allocations are completely not aligned in the beginning (that is not
+ * 2 aligned). This improves unaligned memory access check.
+ *
+ * Also at the beginning we reserve at least SMALL_OBUF_MIN_RESERVE bytes
+ * so that runtime code path with ASAN implementation is similar to regular
+ * one. The overreserve is not very large comparing to regular one to reduce
+ * memory waste.
+ */
+struct obuf {
+	/** For compatibility with existing code only. */
+	struct slab_cache *slabc;
+	/**
+	 * List of iovec vector. NULL terminated. Allocated at creation
+	 * and does not reallocated. Size is SMALL_OBUF_IOV_MAX + 1.
+	 */
+	struct iovec *iov;
```

Comment:
Let's try to embed the array instead of allocating it with `malloc` now, when `SMALL_OBUF_IOV_MAX` is 128.

Reply:
Done.

<details>
<summary>diff</summary>

```diff
diff --git a/include/small/obuf_asan.h b/include/small/obuf_asan.h
index 85d46ea..a84efe7 100644
--- a/include/small/obuf_asan.h
+++ b/include/small/obuf_asan.h
@@ -92,11 +92,8 @@ static_assert(SMALL_OBUF_IOV_MAX <= IOV_MAX,
 struct obuf {
 	/** For compatibility with existing code only. */
 	struct slab_cache *slabc;
-	/**
-	 * List of iovec vector. NULL terminated. Allocated at creation
-	 * and does not reallocated. Size is SMALL_OBUF_IOV_MAX + 1.
-	 */
-	struct iovec *iov;
+	/** List of iovec vectors. NULL terminated. */
+	struct iovec iov[SMALL_OBUF_IOV_MAX + 1];
 	/**
 	 * Vectors starting from SMALL_OBUF_IOV_CHECKED_SIZE index are
 	 * exponentially growing with factor 2 starting from this size.
@@ -128,7 +125,7 @@ obuf_create(struct obuf *buf, struct slab_cache *slabc, size_t start_capacity);
 static inline bool
 obuf_is_initialized(const struct obuf *buf)
 {
-	return buf->iov != NULL;
+	return buf->slabc != NULL;
 }
 
 static inline size_t
diff --git a/small/obuf_asan.c b/small/obuf_asan.c
index 9dac16c..26a8c0d 100644
--- a/small/obuf_asan.c
+++ b/small/obuf_asan.c
@@ -38,9 +38,7 @@ obuf_create(struct obuf *buf, struct slab_cache *slabc, size_t start_capacity)
 {
 	buf->slabc = slabc;
 	buf->start_capacity = start_capacity;
-	size_t size = (SMALL_OBUF_IOV_MAX + 1) * sizeof(*buf->iov);
-	buf->iov = small_xmalloc(size);
-	memset(buf->iov, 0, size);
+	memset(buf->iov, 0, lengthof(buf->iov) * sizeof(buf->iov[0]));
 	memset(buf->capacity, 0,
 	       lengthof(buf->capacity) * sizeof(buf->capacity[0]));
 	buf->pos = 0;
@@ -170,7 +168,6 @@ obuf_destroy(struct obuf *buf)
 	struct obuf_svp svp;
 	obuf_svp_reset(&svp);
 	obuf_rollback_to_svp(buf, &svp);
-	free(buf->iov);
 	/* Safety and also makes mempool_is_initialized work. */
 	memset(buf, 0, sizeof(*buf));
 }

```

</details>